ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/authorization/ApiServiceAuthorization.scala

```diff
@@ -0,0 +1,111 @@
+    val pw = new PrintWriter(sw, true)
+    t.printStackTrace(pw)
+    pw.flush()
+    sw.flush()
+    sw.toString
+  }
+
+  /** Checks whether the given claims give access to the request.
+    *
+    * @param check Given a set of [[Claims]], returns whether the current method is authorized.
+    *              Typically, the caller compares the request parameters with the given claims.
+    *
+    * @return   A [[StatusRuntimeException]] with code `PERMISSION_DENIED` if the request is not authorized,
+    *           or Unit if the request is authorized.
+    */
+  def requireClaims(check: Claims => Authorization): Either[StatusRuntimeException, Unit] = {
+    val claims = AuthorizationInterceptor.contextKeyClaim.get()
+    if (claims == null) {
+      // The current context does not contain any claims.
+      // Most likely causes:
+      // - The API server does not use the [[AuthorizationInterceptor]] which is responsible for creating the context
+      // - This function is called from a thread different from the one used to handle the gRPC call,
+      //   such as during a `Future.map`. See also [[ApiServiceAuthorization.withContext]].
+      logger.error(s"No context key at: ${getStackTrace(new Exception)}")
+      Left(internal("Context.key.get returned no valid Claims."))
+    } else {
+      check(claims) match {
+        case Authorized =>
+          Right(())
+        case NotAuthorized(reason) =>
+          // Be careful with logging errors in order to not leak secrets through log messages
+          // logger.warn(s"Rejecting authorization with claims $claims")
+          Left(permissionDenied(reason))
+      }
+    }
+  }
+
+  /** Executes the given block with the given gRPC context as the current context.
+    *
+    * Usage:
+    * ```
+    * // Capture the gRPC context in a gRPC thread
+    * val ctx = Context.current
+    *
+    * // Function bar uses ApiServiceAuthorization.requireClaims,
+    * // but is called from a non-gRPC thread.
+    * // Wrap it in withContext, so that the gRPC context is available in bar()
+    * val asyncResult: Future[T] = foo()
+    *   .flatMap(ApiServiceAuthorization.withContext(ctx)(bar))
+    * ```
+    */
+  def withContext[T](context: Context)(block: => T): T = {
+    val previous = context.attach()
+    try block
+    finally context.detach(previous)
+  }
+
+  def requirePublicClaims(): Either[StatusRuntimeException, Unit] = {
+    requireClaims(claims => if (claims.isPublic) Authorized else NotAuthorized())
+  }
+
+  def requireAdminClaims(): Either[StatusRuntimeException, Unit] = {
+    requireClaims(claims => if (claims.isAdmin) Authorized else NotAuthorized())
+  }
+
+  def requireClaimsForAllParties(parties: Set[String]): Either[StatusRuntimeException, Unit] = {
+    requireClaims(
+      claims => if (parties.forall(p => claims.canActAs(p))) Authorized else NotAuthorized())
+  }
+
+  def requireClaimsForParty(party: Option[String]): Either[StatusRuntimeException, Unit] = {
```

Comment:
```
scala> None.forall { _ => false }
res1: Boolean = true
```
This would mean that a command submission without a party would go through authorization and be caught by validation instead. It doesn't seem unreasonable so that the concerns are clearly divided, but I thought it would make sense to raise this point for discussion.

Reply:
Good point. I think producing a validation error and not an authorization error for empty party lists is the right thing, so I have left the implementation as is and added a comment.

Let me know if you feel the implementation should change.

Reply:
Cool, thanks for the clarification. Marking as resolved.